File: infraestructura/persistencia/repositorios/SupervisorRepositorio.py
import { PrismaClient, Supervisor as PrismaSupervisor } from '@prisma/client';
import { Supervisor } from '../modelo/Supervisor';
import { Equipo } from '../../monitoreo/Equipo';
import { ISupervisorRepositorio } from './ISupervisorRepositorio';

// Clase de error personalizada para excepciones del repositorio de Supervisor
class SupervisorRepositoryError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'SupervisorRepositoryError';
  }
}

/**
 * Implementación del repositorio para gestionar entidades Supervisor en la base de datos.
 * Cumple con la interfaz ISupervisorRepositorio y usa Prisma para operaciones de persistencia.
 */
export class SupervisorRepositorio implements ISupervisorRepositorio {
  private readonly prisma: PrismaClient;

  constructor(prisma: PrismaClient = new PrismaClient()) {
    this.prisma = prisma;
  }

  /**
   * Agrega un nuevo Supervisor a la base de datos.
   * @param supervisor Objeto Supervisor a persistir
   * @throws SupervisorRepositoryError si la creación falla
   */
  async adicionar(supervisor: Supervisor): Promise<void> {
    try {
      await this.prisma.supervisor.create({
        data: {
          id: supervisor.id,
          nombre: supervisor.nombre,
          apellido: supervisor.apellido,
          equipoId: supervisor.equipoAsignado.id,
        },
      });
    } catch (error) {
      throw new SupervisorRepositoryError(`No se pudo adicionar el Supervisor: ${error instanceof Error ? error.message : 'Error desconocido'}`);
    }
  }

  /**
   * Elimina un Supervisor de la base de datos.
   * @param supervisor Objeto Supervisor a eliminar
   * @throws SupervisorRepositoryError si no se encuentra el Supervisor o la eliminación falla
   */
  async eliminar(supervisor: Supervisor): Promise<void> {
    try {
      const existingSupervisor = await this.prisma.supervisor.findUnique({
        where: { id: supervisor.id },
      });
      if (!existingSupervisor) {
        throw new SupervisorRepositoryError(`Supervisor con ID ${supervisor.id} no encontrado`);
      }
      await this.prisma.supervisor.delete({
        where: { id: supervisor.id },
      });
    } catch (error) {
      throw new SupervisorRepositoryError(`No se pudo eliminar el Supervisor con ID ${supervisor.id}: ${error instanceof Error ? error.message : 'Error desconocido'}`);
    }
  }

  /**
   * Modifica un Supervisor existente en la base de datos.
   * @param supervisor Objeto Supervisor con los datos actualizados
   * @throws SupervisorRepositoryError si no se encuentra el Supervisor o la actualización falla
   */
  async modificar(supervisor: Supervisor): Promise<void> {
    try {
      const existingSupervisor = await this.prisma.supervisor.findUnique({
        where: { id: supervisor.id },
      });
      if (!existingSupervisor) {
        throw new SupervisorRepositoryError(`Supervisor con ID ${supervisor.id} no encontrado`);
      }
      await this.prisma.supervisor.update({
        where: { id: supervisor.id },
        data: {
          nombre: supervisor.nombre,
          apellido: supervisor.apellido,
          equipoId: supervisor.equipoAsignado.id,
        },
      });
    } catch (error) {
      throw new SupervisorRepositoryError(`No se pudo modificar el Supervisor con ID ${supervisor.id}: ${error instanceof Error ? error.message : 'Error desconocido'}`);
    }
  }

  /**
   * Busca un Supervisor por su ID.
   * @param idSupervisor El ID del Supervisor a buscar
   * @returns Objeto Supervisor del dominio o null si no se encuentra
   * @throws SupervisorRepositoryError si la operación falla
   */
  async buscar(idSupervisor: string): Promise<Supervisor | null> {
    try {
      const prismaSupervisor = await this.prisma.supervisor.findUnique({
        where: { id: idSupervisor },
        include: { equipoAsignado: true },
      });
      if (!prismaSupervisor || !prismaSupervisor.equipoAsignado) {
        return null;
      }
      return this.toDomainModel(prismaSupervisor, this.mapToEquipo(prismaSupervisor.equipoAsignado));
    } catch (error) {
      throw new SupervisorRepositoryError(`No se pudo buscar el Supervisor con ID ${idSupervisor}: ${error instanceof Error ? error.message : 'Error desconocido'}`);
    }
  }

  /**
   * Mapea un modelo Prisma Supervisor a un modelo de dominio Supervisor.
   * @param prismaSupervisor El modelo Supervisor de Prisma
   * @param equipo El objeto Equipo del dominio
   * @returns Objeto Supervisor del dominio
   */
  private toDomainModel(prismaSupervisor: PrismaSupervisor, equipo: Equipo): Supervisor {
    return new Supervisor(
      prismaSupervisor.id,
      prismaSupervisor.nombre,
      prismaSupervisor.apellido,
      equipo
    );
  }

  /**
   * Mapea un modelo Prisma Equipo a un modelo de dominio Equipo.
   * @param prismaEquipo El modelo Equipo de Prisma
   * @returns Objeto Equipo del dominio
   */
  private mapToEquipo(prismaEquipo: any): Equipo {
    // Asumiendo que Equipo tiene un constructor con id y nombre
    return new Equipo(prismaEquipo.id, prismaEquipo.nombre);
  }

  /**
   * Cierra la conexión del cliente Prisma.
   * Debe llamarse cuando el repositorio ya no sea necesario.
   */
  async disconnect(): Promise<void> {
    await this.prisma.$disconnect();
  }
}

File: infraestructura/persistencia/repositorios/SupervisorRepositorio.ts
import { PrismaClient } from '../generated/prisma';
import { Equipo } from '../../../Dominio/monitoreo/Equipo';
import { EstadoEquipo as EstadoEquipoDominio } from '../../../Dominio/monitoreo/EstadoEquipo';
import { Supervisor } from '../../../Dominio/usuarios/modelo/supervisor';
import { Operador } from '../../../Dominio/usuarios/modelo/operador';
import { ISupervisorRepositorio } from '../../../Dominio/usuarios/InteracesRepositorio/iSupervisorRepositorio';

const prisma = new PrismaClient();

export class SupervisorRepositorio implements ISupervisorRepositorio {
  async crear(supervisor: Supervisor): Promise<void> {
    await prisma.persona.create({
      data: {
        id: supervisor.id,
        nombre: supervisor.nombre,
        apellido: supervisor.apellido  
      }
    })
    await prisma.supervisor.create({
      data: {
        id: supervisor.id,
        equipos: {
          // Asociacion de los equipos ya existentes al supervisor utilizando la relación many-to-many.
          // uso de `connect` para vincular los objetos `Equipo` que ya existen en la base de datos
          connect: supervisor.equipoAsignados.map(equipo => ({ id: equipo.id }))
        }
      }
    })
  }
  
  async eliminar(id: string): Promise<void> {
    await prisma.supervisor.delete({
      where: { id }
    })
  }
  
  async actualizar(supervisor: Supervisor): Promise<void> {
    await prisma.supervisor.update({
      where: { id: supervisor.id },
      data: {
        persona: {
          update: {
            nombre: supervisor.nombre,
            apellido: supervisor.apellido,
          }
        }
      }
    });
  }
    
  async obtenerPorId(id: string): Promise<Supervisor | null> {
    const supervisorPrisma = await prisma.supervisor.findUnique({
      where: { id },
      include: {
        persona: true,
        equipos: {
          include: {
            operador: {
              include: {
                persona: true
              }
            }
          }
        }
      }
    });

    if (!supervisorPrisma) return null;

    const equipos: Equipo[] = [];

    for (const equipoPrisma of supervisorPrisma.equipos) {
      const operadorPrisma = equipoPrisma.operador;
      // problematica circular
      // Crea un Equipo sin operador primero
      const equipo = new Equipo(
        equipoPrisma.id,
        equipoPrisma.codigo,
        equipoPrisma.tipo,
        equipoPrisma.nivelCombustible,
        equipoPrisma.horasOperacion
      );

      // Crea al operador con equipo asignado
      const operador = new Operador(
        operadorPrisma.id,
        operadorPrisma.persona.nombre,
        operadorPrisma.persona.apellido,
        operadorPrisma.licencia,
        equipo
      );

      // Asigna el operador al equipo
      equipo.operadorAsignado = operador;

      equipos.push(equipo);
    }

    const supervisor = new Supervisor(
      supervisorPrisma.id,
      supervisorPrisma.persona.nombre,
      supervisorPrisma.persona.apellido,
      equipos
    );

    return supervisor;
  }


  async obtenerTodos(): Promise<Supervisor[]> {
    const rows = await prisma.supervisor.findMany({
      include: {
        persona: true,
        equipos: {
          include: {
            operador: {
              include: { persona: true }
            }
          }
        }
      }
    });

    const supervisores = rows.map(row => {
      const equipos: Equipo[] = row.equipos.map(eq => {
        let operador: Operador | undefined;

        if (eq.operador) {
          operador = new Operador(
            eq.operador.id,
            eq.operador.persona.nombre,
            eq.operador.persona.apellido,
            eq.operador.licencia,
            undefined as any // Puedes pasar supervisor si lo necesitas
          );
        }

        const equipo = new Equipo(
          eq.id,
          eq.codigo,
          eq.tipo,
          eq.nivelCombustible,
          eq.horasOperacion,
          operador
        );

        if (operador) {
          equipo.operadorAsignado = operador;
        }

        equipo.estado = EstadoEquipoDominio[eq.estado as keyof typeof EstadoEquipoDominio];
        return equipo;
      });

      return new Supervisor(
        row.id,
        row.persona.nombre,
        row.persona.apellido,
        equipos
      );
    });

    return supervisores;
  }
    
  async buscarPorNombre(nombre: string): Promise<Supervisor[]> {
    const rows = await prisma.supervisor.findMany({
      where: {
        persona: {
          nombre: {
            contains: nombre,
            mode: 'insensitive', // búsqueda sin distinguir mayúsculas/minúsculas
          },
        },
      },
      include: {
        persona: true,
        equipos: {
          include: {
            operador: {
              include: { persona: true }
            }
          }
        }
      }
    });

    return rows.map(row => {
      const equipos: Equipo[] = row.equipos.map(eq => {
        let operador: Operador | undefined;

        if (eq.operador) {
          operador = new Operador(
            eq.operador.id,
            eq.operador.persona.nombre,
            eq.operador.persona.apellido,
            eq.operador.licencia,
            undefined as any
          );
        }

        const equipo = new Equipo(
          eq.id,
          eq.codigo,
          eq.tipo,
          eq.nivelCombustible,
          eq.horasOperacion,
          operador
        );

        if (operador) {
          equipo.operadorAsignado = operador;
        }

        // Conversión segura del enum
        equipo.estado = EstadoEquipoDominio[eq.estado as keyof typeof EstadoEquipoDominio];

        return equipo;
      });

      return new Supervisor(
        row.id,
        row.persona.nombre,
        row.persona.apellido,
        equipos
      );
    });
  }

  
  async buscarPorApellido(apellido: string): Promise<Supervisor[]> {
    const rows = await prisma.supervisor.findMany({
      where: {
        persona: {
          nombre: {
            contains: apellido,
            mode: 'insensitive', // búsqueda sin distinguir mayúsculas/minúsculas
          },
        },
      },
      include: {
        persona: true,
        equipos: {
          include: {
            operador: {
              include: { persona: true }
            }
          }
        }
      }
    });

    return rows.map(row => {
      const equipos: Equipo[] = row.equipos.map(eq => {
        let operador: Operador | undefined;

        if (eq.operador) {
          operador = new Operador(
            eq.operador.id,
            eq.operador.persona.nombre,
            eq.operador.persona.apellido,
            eq.operador.licencia,
            undefined as any
          );
        }

        const equipo = new Equipo(
          eq.id,
          eq.codigo,
          eq.tipo,
          eq.nivelCombustible,
          eq.horasOperacion,
          operador
        );

        if (operador) {
          equipo.operadorAsignado = operador;
        }

        // Conversión segura del enum
        equipo.estado = EstadoEquipoDominio[eq.estado as keyof typeof EstadoEquipoDominio];

        return equipo;
      });

      return new Supervisor(
        row.id,
        row.persona.nombre,
        row.persona.apellido,
        equipos
      );
    });
  }

  async listar(): Promise<Supervisor[]> {
    const rows = await prisma.supervisor.findMany({
      include: {
        persona: true,
        equipos: {
          include: {
            operador: {
              include: {
                persona: true
              }
            }
          }
        }
      }
    });

    return rows.map(row => {
      const equipos: Equipo[] = row.equipos.map(eq => {
        let operador: Operador | undefined;

        if (eq.operador) {
          operador = new Operador(
            eq.operador.id,
            eq.operador.persona.nombre,
            eq.operador.persona.apellido,
            eq.operador.licencia,
            undefined as any // Aquí puedes inyectar un repositorio si es necesario
          );
        }

        const equipo = new Equipo(
          eq.id,
          eq.codigo,
          eq.tipo,
          eq.nivelCombustible,
          eq.horasOperacion,
          operador
        );

        equipo.estado = EstadoEquipoDominio[eq.estado as keyof typeof EstadoEquipoDominio];

        return equipo;
      });

      return new Supervisor(
        row.id,
        row.persona.nombre,
        row.persona.apellido,
        equipos
      );
    });
  }
}